app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -46,11 +46,18 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     yum localinstall $CONSUMER_RPM -y
     yum install subscription-manager -y
     subscription-manager register --org=<%= @organization.label %> --activationkey="sample-key"
-
-    rm -f $CONSUMER_RPM
 else
     create_host_curl
 fi
 <% else -%>
 create_host_curl
 <% end -%>
+
+if ! [ $? = 0 ]; then
+    echo ""
+    echo "Host registration failed"
+    exit 1
+fi
+
+TOKEN=$(curl -s -X GET "<%= foreman_server_url %>/api/hosts?search=name+%3D+$(hostname --fqdn)" <%= headers.join(' ') %> | grep -Po '(?<=\"registration_token\":")([^\"]+)')
```

Comment:
For the hardcored paths, I updated https://projects.theforeman.org/issues/30742, I don't want to change it here since there is another  https://github.com/theforeman/foreman/pull/7928 for URL and it could get messy.

But still, if users change the template (let's say add another api calls), they won't have access to path helpers, right? So they will put the url into the template anyway. 

Is it really worth it to have another variables just for paths, what is the chance of changing API paths?

Reply:
Adding cc @ekohl, we have same discussion here: https://github.com/theforeman/foreman/pull/7928#discussion_r487215543

Reply:
> Is it really worth it to have another variables just for paths, what is the chance of changing API paths?

In the case of Foreman Proxy being in between: 100%. That is not something that should be done in templates since it's not something users should even think about. URL construction is messy and error prone.

Another is that the Smart Proxy will need to be authenticated for the Katello side of it. Only then can you replace some arguments. Having all that logic in the template is IMHO a bad idea. It's part of the core functionality, not something an end user should modify.

Reply:
(I updated the comment meanwhile, so maybe you didn't see my changes.)

> In the case of Foreman Proxy being in between: 100%

Is it a concern for the Registration module, where we don't pass full URL?
https://github.com/theforeman/smart-proxy/pull/768/files#diff-45f33bea51f0f06fd02907550e854ae2R7
``` ruby
    def get_template(request)
      params = request.params
      params[:url] = request.env['REQUEST_URI']&.split('/register')&.first
      proxy_req = request_factory.create_get '/register', params, headers(request)
      send_request(proxy_req).body
    end
```

> Another is that the Smart Proxy will need to be authenticated for the Katello side of it. Only then can you replace some arguments.

How does it related to the Registration template?

> Having all that logic in the template is IMHO a bad idea.

What logic exactly? I agree that base URL should be defined in the controller, like this:
```ruby
@url = params[:url] || foreman_server_url // Just an example
```
but defining routes with paths should be done in the templates.

Reply:
> How does it related to the Registration template?

For the Katello side of it you need a Smart Proxy instance. I've explained this in other PRs, but because there's so many of them I think it's getting lost.

Another is that right now you're *developing* in this template. At some point in 2.3 we'll branch. Maybe we find a bug somewhere in the URL logic. Users may have copied 2.3 and they don't receive the bugfix. By providing the URLs in a context, the user just needs to remember to use the correct variable.

This is in line with `foreman_url`:
https://github.com/theforeman/foreman/blob/c1bc8cd6edc8b44b033ccc524989fb6c5eceacfe/lib/foreman/foreman_url_renderer.rb#L11-L36

Note that there we also properly deal with GET parameters which is always painful in templates.

That particular method may not be entirely correct for this, but that's how we have traditionally dealt with this.

Reply:
> For the Katello side of it you need a Smart Proxy instance. I've explained this in other PRs, but because there's so many of them I think it's getting lost.

Is it this https://github.com/Katello/katello/pull/8938 ?

I've been thinking about other cases:
- If GRT is rendered via proxy, is there a helper for generating proxy URLs like ther is for Foreman - `foreman_url`?
If not, what to do in that case?
- User is updating template and adding his own commands, for example adding facts upload via API:
```
curl -X POST <% @ foreman_url('facts_action') '%>
```
What to do if he will render the template via proxy? He can't use `foreman_url`.